hivelocity/instances_v2.go

```diff
@@ -82,60 +85,93 @@ func getHivelocityDeviceIDFromNode(node *corev1.Node) (int32, error) {
+		if errors.Is(err, client.ErrNoSuchDevice) {
+			return nil, nil
+		}
+		if err != nil {
+			return nil, fmt.Errorf(
+				"[InstanceExists] GetBareMetalDevice() failed. node %q, deviceID %d: %w",
+				node.GetName(),
+				deviceID,
+				err,
+			)
+		}
+		return device, nil
+	} else {
+		devices, err := i2.client.ListDevice(ctx)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"[InstanceExists] ListDevice() failed. node %q: %w",
+				node.GetName(),
+				err,
+			)
+		}
+
+		for _, device := range devices {
+			name, err := hvutils.GetMachineNameFromTags(device.Tags)
+			if err != nil {
+				return nil, fmt.Errorf(
+					"InstanceExists(): GetMachineNameFromTags() failed. node %q: %w",
+					node.GetName(),
+					err,
+				)
+			}
+
+			if name == node.GetName() {
+				return &device, nil
+			}
+		}
+
+		return nil, client.ErrNoSuchDevice
+	}
+}
+
 // InstanceExists returns true if the instance for the given node exists according to the cloud provider.
 // Use the node.name or node.spec.providerID field to find the node in the cloud provider.
 // Implements cloudprovider.InstancesV2.InstanceExists.
 func (i2 *HVInstancesV2) InstanceExists(ctx context.Context, node *corev1.Node) (bool, error) {
+	const op = "hivelocity/instancesv2.InstanceExists"
+
 	if node == nil {
 		return false, errNodeIsNil
 	}
-	deviceID, err := getHivelocityDeviceIDFromNode(node)
-	if err != nil {
-		return false, fmt.Errorf(
-			"[InstanceExists] getHivelocityDeviceIDFromNode() failed. node %q: %w",
-			node.GetName(),
-			err,
-		)
-	}
-	_, err = i2.client.GetBareMetalDevice(ctx, deviceID)
-	if errors.Is(err, client.ErrNoSuchDevice) {
-		return false, nil
-	}
+
+	device, err := i2.lookUpDevice(ctx, node)
 	if err != nil {
-		return false, fmt.Errorf(
-			"[InstanceExists] GetBareMetalDevice() failed. node %q, deviceID %d: %w",
-			node.GetName(),
-			deviceID,
-			err,
-		)
+		return false, fmt.Errorf("%s: %w", op, err)
 	}
-	return true, nil
+
+	return device != nil, nil
 }
 
 // InstanceShutdown returns true if the instance is shutdown according to the cloud provider.
 // Use the node.name or node.spec.providerID field to find the node in the cloud provider.
 // Implements cloudprovider.InstancesV2.InstanceShutdown.
 func (i2 *HVInstancesV2) InstanceShutdown(ctx context.Context, node *corev1.Node) (bool, error) {
+	const op = "hivelocity/instancesv2.InstanceShutdown"
+
 	if node == nil {
 		return false, errNodeIsNil
 	}
-	deviceID, err := getHivelocityDeviceIDFromNode(node)
-	if err != nil {
-		return false, fmt.Errorf(
-			"[InstanceShutdown] getHivelocityDeviceIDFromNode() failed. node %q: %w",
-			node.GetName(),
-			err,
-		)
-	}
-	device, err := i2.client.GetBareMetalDevice(ctx, deviceID)
+
+	device, err := i2.lookUpDevice(ctx, node)
 	if err != nil {
-		return false, fmt.Errorf(
-			"[InstanceShutdown] GetBareMetalDevice() failed. deviceID %d, node %q: %w",
-			deviceID,
-			node.GetName(),
-			err,
-		)
+		return false, fmt.Errorf("%s: %w", op, err)
 	}
+
```

Comment:
what if device is nil? -> should be an error I guess. re-check in cloud-provider interface for more infos and Hetzner CCM for another resource